### ISSUE TITLE
[JENKINS-22569] git plugin checks out HEAD instead of $GERRIT_REFSPEC for manual triggers

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/gerrit-trigger-plugin#670]
<p>As described in the documentation, Gerrit allows you to start a git job manually <a href="https://wiki.jenkins-ci.org/display/JENKINS/Gerrit+Trigger#GerritTrigger-Using%22BuildNow%22" class="external-link" target="_blank" rel="nofollow noopener">by specifying the <tt>GERRIT_REFSPEC</tt> parameter</a>.<br/>
In my setup the value of this parameter is ignored and gerrit checks out the remote <tt>HEAD</tt> instead of the specified refspec.</p>

<p>I tried to trace down the issue and found out that the <em>Gerrit merge</em> strategy will select the <tt>FETCH_HEAD</tt> to start the build from.  <br/>
The problem is however that the git-client not only fetches the <tt>$GERRIT_REFSPEC</tt> but also fetches an empty refspec, so on the command line, it executes:</p>
<blockquote><p>git fetch --tags --progress <url of repo> <b>""</b> $GERRIT_REFSPEC</p></blockquote>

<p>The empty parameter will be translated on the remote site to the <tt>HEAD</tt> and this is where the <tt>FETCH_HEAD</tt> will be put on.</p>

<p>So the build will start from the wrong git version.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=steven_aerts">steven_aerts</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22569" target="_blank">git plugin checks out HEAD instead of $GERRIT_REFSPEC for manual triggers</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rsandell.png" title="rsandell's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=rsandell">rsandell</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Major
<li><b>component(s)</b>: gerrit-trigger-plugin, git-client-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-10-14T21:21:16+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>As described in the documentation, Gerrit allows you to start a git job manually <a href="https://wiki.jenkins-ci.org/display/JENKINS/Gerrit+Trigger#GerritTrigger-Using%22BuildNow%22" class="external-link" target="_blank" rel="nofollow noopener">by specifying the <tt>GERRIT_REFSPEC</tt> parameter</a>.
In my setup the value of this parameter is ignored and gerrit checks out the remote <tt>HEAD</tt> instead of the specified refspec.</p>

<p>I tried to trace down the issue and found out that the <em>Gerrit merge</em> strategy will select the <tt>FETCH_HEAD</tt> to start the build from.  
The problem is however that the git-client not only fetches the <tt>$GERRIT_REFSPEC</tt> but also fetches an empty refspec, so on the command line, it executes:</p>
<blockquote><p>git fetch --tags --progress &lt;url of repo&gt; <b>""</b> $GERRIT_REFSPEC</p></blockquote>

<p>The empty parameter will be translated on the remote site to the <tt>HEAD</tt> and this is where the <tt>FETCH_HEAD</tt> will be put on.</p>

<p>So the build will start from the wrong git version.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>ubuntu 12.04</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>steven_aerts:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/22569?focusedId=198416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-198416">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22569?focusedId=198416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-198416">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I tried to find out with the debugger where this empty refspec is coming from (gerrit-trigger, git-client-plugin, git-plugin or jgit).  But I was unable to do so.</p>

<p>To unblock myself, I altered the git-client-plugin to ignore empty refspecs with the following hack:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">--- a/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
+++ b/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
@@ -244,8 +244,12 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>CliGitAPIImpl <span class="code-keyword">extends</span> LegacyCompatibleGitAPIImpl {
 
                 <span class="code-keyword">if</span> (refspecs != <span class="code-keyword">null</span>)
                     <span class="code-keyword">for</span> (RefSpec rs: refspecs)
-                        <span class="code-keyword">if</span> (rs != <span class="code-keyword">null</span>)
+                        <span class="code-keyword">if</span> (rs != <span class="code-keyword">null</span>) {
+                          <span class="code-object">String</span> refString = rs.toString();
+                          <span class="code-keyword">if</span>(!StringUtils.isBlank(refString)) {
                             args.add(rs.toString());
+                          }
+                        }
 
                 <span class="code-keyword">if</span> (prune) args.add(<span class="code-quote">"--prune"</span>);
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>I tried to find out with the debugger where this empty refspec is coming from (gerrit-trigger, git-client-plugin, git-plugin or jgit).  But I was unable to do so.</p>

<p>To unblock myself, I altered the git-client-plugin to ignore empty refspecs with the following hack:</p>


<pre>
--- a/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
+++ b/src/main/java/org/jenkinsci/plugins/gitclient/CliGitAPIImpl.java
@@ -244,8 +244,12 @@ <span class="code-keyword">public</span> <span class="code-keyword">class </span>CliGitAPIImpl <span class="code-keyword">extends</span> LegacyCompatibleGitAPIImpl {
 
 <span class="code-keyword">if</span> (refspecs != <span class="code-keyword">null</span>)
     <span class="code-keyword">for</span> (RefSpec rs: refspecs)
-<span class="code-keyword">if</span> (rs != <span class="code-keyword">null</span>)
+<span class="code-keyword">if</span> (rs != <span class="code-keyword">null</span>) {
+  <span class="code-object">String</span> refString = rs.toString();
+  <span class="code-keyword">if</span>(!StringUtils.isBlank(refString)) {
     args.add(rs.toString());
+  }
+}
 
 <span class="code-keyword">if</span> (prune) args.add(<span class="code-quote">"--prune"</span>);
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>steven_aerts:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/22569?focusedId=198418&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-198418">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22569?focusedId=198418&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-198418">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It might by possible that <a href="https://issue-redirect.jenkins.io/issue/15313" title="Trigger on Updated Change does not fetch the last commit " class="issue-link" data-issue-key="JENKINS-15313">JENKINS-15313</a> is a duplicate of this issue.</p>
</pre>
</details>

</ul>
</details>

<p>It might by possible that <a href="https://issue-redirect.jenkins.io/issue/15313" title="Trigger on Updated Change does not fetch the last commit " class="issue-link" data-issue-key="JENKINS-15313">JENKINS-15313</a> is a duplicate of this issue.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/markewaite.png" title="markewaite's avatar" /> markewaite:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/22569?focusedId=316895&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316895">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-22569?focusedId=316895&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-316895">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe this was resolved with the 2.5.3 release of the git plugin which added the "Honor refspec on initial clone" option to the "Advanced clone options".</p>

<p>The git plugin original behavior was to clone a broad refspec initially, then honor the user specified refspec for later operations.  That was a strange default, and there were a number of users that depended on that strange default.  When "Honor refspec" is enabled in the "Advanced clone options", then the initial clone will use the refspecs as provided by the user, rather than initially using the default refspec, then switching to the user provided refspec for updates.</p>
</pre>
</details>

</ul>
</details>

<p>I believe this was resolved with the 2.5.3 release of the git plugin which added the "Honor refspec on initial clone" option to the "Advanced clone options".</p>

<p>The git plugin original behavior was to clone a broad refspec initially, then honor the user specified refspec for later operations.  That was a strange default, and there were a number of users that depended on that strange default.  When "Honor refspec" is enabled in the "Advanced clone options", then the initial clone will use the refspecs as provided by the user, rather than initially using the default refspec, then switching to the user provided refspec for updates.</p>
